File: app/schemas/appointments.py
"""Schemas para citas/agenda."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.appointments import AppointmentStatus


class AppointmentBase(BaseModel):
    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_phone: Optional[str] = Field(None, max_length=64)
    vehicle_plates: Optional[str] = Field(None, max_length=50)
    vehicle_model_id: Optional[str] = None
    scheduled_at: datetime
    duration_minutes: int = Field(60, ge=15, le=600)
    service_type: str = Field(..., min_length=1, max_length=120)
    notes: Optional[str] = None


class AppointmentCreate(AppointmentBase):
    branch_id: Optional[str] = None


class AppointmentUpdate(BaseModel):
    customer_name: Optional[str] = None
    customer_phone: Optional[str] = None
    vehicle_plates: Optional[str] = None
    vehicle_model_id: Optional[str] = None
    scheduled_at: Optional[datetime] = None
    duration_minutes: Optional[int] = None
    service_type: Optional[str] = None
    notes: Optional[str] = None


class AppointmentRead(BaseModel):
    id: str
    branch_id: str
    customer_name: str
    customer_phone: Optional[str] = None
    vehicle_plates: Optional[str] = None
    vehicle_model_id: Optional[str] = None
    scheduled_at: datetime
    duration_minutes: int
    service_type: str
    notes: Optional[str] = None
    status: AppointmentStatus
    work_order_id: Optional[str] = None
    cancel_reason: Optional[str] = None
    created_by_id: Optional[str] = None
    arrived_at: Optional[datetime] = None
    converted_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class AppointmentCancelRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)


class AppointmentConvertRequest(BaseModel):
    """Datos requeridos para convertir cita en OS.

    Si el vehículo ya existe (por placas) se reutiliza; si no, se crea.
    """

    service_id: str = Field(..., min_length=1)
    vehicle_id: Optional[str] = None  # si ya hay vehículo en sistema
    assigned_mechanic_id: Optional[str] = None
    notes: Optional[str] = None


class AppointmentConvertResponse(BaseModel):
    appointment: AppointmentRead
    work_order: dict
    whatsapp_link: Optional[str] = None

File: frontend/src/api/endpoints/appointments.ts
/** Client de /api/v1/appointments. */
import api from "@/api/client";

export type AppointmentStatus =
  | "scheduled"
  | "arrived"
  | "cancelled"
  | "no_show"
  | "converted_to_wo";

export interface AppointmentRead {
  id: string;
  branch_id: string;
  customer_name: string;
  customer_phone: string | null;
  vehicle_plates: string | null;
  vehicle_model_id: string | null;
  scheduled_at: string;
  duration_minutes: number;
  service_type: string;
  notes: string | null;
  status: AppointmentStatus;
  work_order_id: string | null;
  cancel_reason: string | null;
  created_by_id: string | null;
  arrived_at: string | null;
  converted_at: string | null;
  cancelled_at: string | null;
  created_at: string;
  updated_at: string | null;
}

export interface AppointmentCreatePayload {
  customer_name: string;
  customer_phone?: string | null;
  vehicle_plates?: string | null;
  vehicle_model_id?: string | null;
  scheduled_at: string;
  duration_minutes?: number;
  service_type: string;
  notes?: string | null;
}

export interface AppointmentConvertResponse {
  appointment: AppointmentRead;
  work_order: {
    id: string;
    order_number: string;
    status: string;
    branch_id: string;
  };
  whatsapp_link: string | null;
}

export interface AppointmentListParams {
  date_from?: string;
  date_to?: string;
  branch_id?: string;
  status?: AppointmentStatus;
}

export const appointmentsApi = {
  list: (params: AppointmentListParams = {}) =>
    api.get<AppointmentRead[]>(`/v1/appointments`, { params }).then((r) => (Array.isArray(r.data) ? r.data : [])),

  create: (payload: AppointmentCreatePayload) =>
    api.post<AppointmentRead>(`/v1/appointments`, payload).then((r) => r.data),

  update: (id: string, payload: Partial<AppointmentCreatePayload>) =>
    api.patch<AppointmentRead>(`/v1/appointments/${id}`, payload).then((r) => r.data),

  remove: (id: string) => api.delete(`/v1/appointments/${id}`).then(() => undefined),

  markArrived: (id: string) =>
    api.post<AppointmentRead>(`/v1/appointments/${id}/arrived`).then((r) => r.data),

  cancel: (id: string, reason: string) =>
    api.post<AppointmentRead>(`/v1/appointments/${id}/cancel`, { reason }).then((r) => r.data),

  convert: (
    id: string,
    payload: { service_id: string; vehicle_id?: string; assigned_mechanic_id?: string; notes?: string },
  ) => api.post<AppointmentConvertResponse>(`/v1/appointments/${id}/convert`, payload).then((r) => r.data),
};
